content/solids/components/voxel-painter.ts

```diff
@@ -0,0 +1,462 @@
+        if (placingStart.z === placingEnd.z) iComponent = 0
+        let jComponent = (iComponent + 1) % 3
+        let kComponent = (jComponent + 1) % 3
+
+        let iStart = Math.min( placingStart.getComponent(iComponent), placingEnd.getComponent(iComponent) )
+        let iLimit = Math.max( placingStart.getComponent(iComponent), placingEnd.getComponent(iComponent) )
+        let jStart = Math.min( placingStart.getComponent(jComponent), placingEnd.getComponent(jComponent) )
+        let jLimit = Math.max( placingStart.getComponent(jComponent), placingEnd.getComponent(jComponent) )
+
+        // log(placingStart,placingEnd, iStart,iLimit,jStart,jLimit)
+        let potentialNewPosition = new THREE.Vector3()
+        for ( let i = iStart; i <= iLimit; i+= 1.)
+          for ( let j = jStart; j <= jLimit; j += 1.)
+          {
+            let k = placingStart.getComponent(kComponent)
+            potentialNewPosition.setComponent(kComponent, k)
+            potentialNewPosition.setComponent(iComponent, i)
+            potentialNewPosition.setComponent(jComponent, j)
+            snapToNearestValidCubeCenterPosition(potentialNewPosition)
+            let alreadyOccupied = false
+            voxels.forEach((voxel)=>{
+              if(voxel.position.equals(potentialNewPosition)) alreadyOccupied = true
+            })
+            if(!alreadyOccupied)
+              Voxel().position.copy(potentialNewPosition)
+          }
+      }
+      
+      mouseControlMode = ""
+      placementVisualizer.scale.setScalar(1.)
+    });
+
+    function getCameraSpaceDirectionInAxisPlane(worldSpaceDirection, axisPlane, target)
+    {
+      target.copy(worldSpaceDirection)
+      camera.worldToLocal(target)
+      target[axisPlane] = 0.
+      target.normalize()
+    }
+
+    function getStepTowardDestination(currentValue,destination)
+    {
+      let distanceFromDestination = destination - currentValue
+      let sign = distanceFromDestination == 0. ? 0. : distanceFromDestination / Math.abs(distanceFromDestination)
+      let speed = .01
+      if(speed > Math.abs(distanceFromDestination))
+        speed = Math.abs(distanceFromDestination)
+      return sign * speed
+    }
+
+    updateFunctions.push(()=>
+    {
+      //camera
+      {
+        if (mouseControlMode === "rotating") {
+          getCameraSpaceDirectionInAxisPlane(raycaster.ray.direction, "y", v1)
+          getCameraSpaceDirectionInAxisPlane(oldRay.direction, "y", v2)
+          let horizontalDelta = v1.angleTo(v2) * (v1.x < v2.x ? 1. : -1.)
+
+          getCameraSpaceDirectionInAxisPlane(raycaster.ray.direction, "x", v1)
+          getCameraSpaceDirectionInAxisPlane(oldRay.direction, "x", v2)
+          let verticalDelta = v1.angleTo(v2) * (v1.y > v2.y ? 1. : -1.)
+
+          camera.rotation.y += horizontalDelta * 40.
+          camera.rotation.x += verticalDelta * 35
+          camera.rotation.x = clamp(camera.rotation.x, -TAU / 4., TAU / 4.)
```

Comment:
I think we should rotate the object, not the camera, to avoid weird lighting effects

Reply:
To avoid weird lighting effects I've attached the light to the camera instead, hope that's ok? To rotate the objects in the scene would entail a bunch more worldToLocal calls, especially for placement, which I find often cause unpleasant bugs.